### ISSUE TITLE
False positive for conflicting implementations

### ISSUE THREAD

**Issue by GoldsteinE** [rust-lang/rust#102731]
I tried this code ([playground]):

```rust
#![feature(try_trait_v2)]

use std::ops::{Try, FromResidual, ControlFlow};

#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord, Hash)]
#[repr(u8)]
enum Bool {
    False = 0,
    True = 1,
}

use Bool::{False, True};

impl Try for Bool {
    type Output = Self;
    type Residual = Self;
    
    fn from_output(output: Self) -> Self {
        output
    }
    
    fn branch(self) -> ControlFlow<Self, Self> {
        match self {
            False => ControlFlow::Break(self),
            True => ControlFlow::Continue(self),
        }
    }
}

impl FromResidual<Self> for Bool {
    fn from_residual(residual: Self) -> Self {
        residual
    }
}

impl<T> FromResidual<Bool> for Option<T> {
    fn from_residual(residual: Bool) -> Self {
        None
    }
}
```

[playground]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5f1c551c59778be5fa911dd8015066ba

I expected to see the code to compile.

Instead this failed to compile with the following error:

```
Compiling playground v0.0.1 (/playground)
error[[E0119]](https://doc.rust-lang.org/nightly/error-index.html#E0119): conflicting implementations of trait `std::ops::FromResidual<Bool>` for type `std::option::Option<_>`
  --> src/lib.rs:36:1
   |
36 | impl<T> FromResidual<Bool> for Option<T> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T> FromResidual for Option<T>;

For more information about this error, try `rustc --explain E0119`.
error: could not compile `playground` due to previous error
```

I don’t think that `FromResidual<Bool> for Option<T>` conflicts with `FromResidual for Option<T>`, since the latter is actually `FromResidual<<Option<T> as Try>::Residual> for Option<T>`, which is `FromResidual<Option<Infallible>> for Option<T>`, and `Bool` is always distinct from `Option<T>`.

### Meta

Reproduces on playground  (1.66.0-nightly (2022-10-05 c97d02cdb5ca5f5e9eff))

**Comment by BGR360** [comment]
Minimal repro:

In `lib.rs`:

```rust
pub trait A {  // (Try)
    type Assoc;
}

pub trait B<T> {}  // (FromResidual)

pub struct Foreign;  // (Option)

impl A for Foreign {
    type Assoc = ();
}

impl B<<Self as A>::Assoc> for Foreign {}
```

In `main.rs`:

```rust
use lib::{Foreign, B};

struct Local;  // (Bool)

impl B<Local> for Foreign {}

fn main() {}
```

```
error[E0119]: conflicting implementations of trait `issue_102731::B<Local>` for type `issue_102731::Foreign`
 --> src/main.rs:5:1
  |
5 | impl B<Local> for Foreign {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `issue_102731`:
          - impl B<<Foreign as A>::Assoc> for Foreign;
```

Compiles fine if code is combined into one crate.

This rules out a few things:
* It's not related to the fact that `FromResidual` is a subtrait of `Try`.
* It's not related to the fact that `FromResidual<T>` has a default value for `T`.
* It's not related to the fact that `Option` is generic.
* It doesn't matter what type is used for `<Foreign as A>::Assoc`.
* `Local` doesn't even need to implement the "supertrait".

And here's another hint: if you change `<Self as A>::Assoc` to just directly say `()` instead, it compiles:

```rust
impl B<()> for Foreign {}
```

Not sure if this is `T-types` or `T-lang`; I'm not hip on the difference yet. But I ask unto either of y'all: is it not okay to allow this? `<lib::Foreign as A>::Assoc` cannot possibly be the same as `Local`, not now nor in the future, because:
1. The implementation of `A for Foreign` comes from the upstream crate.
1. `Local` is defined in a downstream crate

@rustbot label +T-types +T-lang +S-bug-has-mcve

**Comment by lolbinarycat** [comment]
triage: original issue using `FromResidual` is fixed, but the minimal repro still causes an error.

**Comment by lolbinarycat** [comment]
here's a repro in `txtar` format:
```rust
-- issue-102731/Cargo.toml --
[package]
name = "issue-102731"
version = "0.1.0"
edition = "2021"

[dependencies]

-- issue-102731/src/lib.rs --
pub trait A {  // (Try)
    type Assoc;
}

pub trait B<T> {}  // (FromResidual)

pub struct Foreign;  // (Option)

impl A for Foreign {
    type Assoc = ();
}

impl B<<Self as A>::Assoc> for Foreign {}

-- issue-102731/src/main.rs --
use issue_102731::{Foreign, B};

struct Local;  // (Bool)

impl B<Local> for Foreign {}

fn main() {}

```
